File: lang/syntax/token.swift
/**
 * @file token.swift
 * @author Krisna Pranav
 * @brief Token
 * @version 6.0
 * @date 2024-10-20
 *
 * @copyright Copyright (c) 2021-2024 pranaOS Developers, Krisna Pranav
 *
 */

struct Token {
    enum Kind {
        case associated(String)
        case keyword(String)
        case punctuation(String)
    }

    let caseName: String
    let name: String
    let kind: Kind

    init(name: String, _ kind: Kind) {
        self.kind = kind
        self.name = name

        var caseName = name.replacingOccurrences(of: "Token", with: "").lowercaseFirstLetter

        if case .keyword(_) = kind {
            caseName += "Keyword"
        }

        self.caseName = caseName
    }
}

File: libraries/libcpp/token.cpp
/**
 * @file token.cpp
 * @author Krisna Pranav
 * @brief token
 * @version 6.0
 * @date 2025-03-05
 * 
 * @copyright Copyright (c) 2021-2025 pranaOS Developers, Krisna Pranav
 * 
 */

#include "token.h"
#include <mods/string.h>

namespace Cpp
{

    /**
     * @param other 
     * @return true 
     * @return false 
     */
    bool Position::operator<(Position const& other) const
    {
        return line < other.line || (line == other.line && column < other.line);
    }

    /**
     * @return String 
     */
    String Token::to_string() const
    {
        return String::formatted("{} {}:-{}:{}");
    }

    /**
     * @return String 
     */
    String Token::type_as_string() const 
    {
        return type_to_string(m_type);
    }
} // namespace Cpp
